lib/admin/WP_Auth0_Admin_Advanced.php

```diff
@@ -134,294 +134,233 @@ public function render_passwordless_method() {
-    <div class="subelement">
-      <span class="description">
-        <?php echo __( 'This field is the Json that describes the custom signup fields for lock. It should be a valida json and allows the use of functions (for validation). More info', 'wp-auth0' ); ?>
-        <a target="_blank" href="https://auth0.com/docs/libraries/lock/v10/new-features#custom-sign-up-fields"><?php echo __( 'here', 'wp-auth0' ); ?></a>
-
-        <code><pre>[
-  {
-    name: "address",                              // required
-    placeholder: "enter your address",            // required
-    icon: "https://example.com/address_icon.png", // optional
-    prefill: "street 123",                        // optional
-    validator: function(value) {                  // optional
-      // only accept addresses with more than 10 chars
-      return value.length > 10;
-    }
-  },
-  {
-    ... // more fields could be specified
-  }
-]</pre></code>
-      </span>
-    </div>
-    <?php
+  public function render_custom_signup_fields( $args ) {
+    $this->render_textarea_field( $args[ 'label_for' ], $args[ 'opt_name' ] );
+    $this->render_field_description(
+      __( 'Valid JSON for custom signup fields in the Auth0 signup form. ', 'wp-auth0' ) .
+      $this->get_docs_link(
+        'libraries/lock/v11/configuration#additionalsignupfields-array-',
+        __( 'More information and examples', 'wp-auth0' )
+      )
+    );
   }
 
-  public function render_link_auth0_users() {
-    $v = $this->options->get( 'link_auth0_users' );
-
-    echo $this->render_a0_switch( "wpa0_link_auth0_users", "link_auth0_users", 1, ! empty( $v ) );
-?>
-      <div class="subelement">
-        <span class="description"><?php echo __( 'Links accounts with the same e-mail address. It will only occur if both e-mails are previously verified.', 'wp-auth0' ); ?></span>
-      </div>
-    <?php
+  public function render_link_auth0_users( $args ) {
+    $this->render_switch( $args[ 'label_for' ], $args[ 'opt_name' ] );
+    $this->render_field_description(
+      __( 'Links accounts with the same e-mail address (emails must be verified)', 'wp-auth0' )
+    );
   }
 
-  public function render_auto_provisioning() {
-    $v = $this->options->get( 'auto_provisioning' );
-
-    echo $this->render_a0_switch( "wpa0_auto_provisioning", "auto_provisioning", 1, 1 == $v );
-?>
-
-      <div class="subelement">
-        <span class="description"><?php echo __( 'The plugin will automatically add new users if they do not exist in the WordPress database if the signups are enabled (enabling this setting will enable this behaviour when signups are disabled).', 'wp-auth0' ); ?></span>
-        </div>
-    <?php
+  public function render_auto_provisioning( $args ) {
+    $this->render_switch( $args[ 'label_for' ], $args[ 'opt_name' ] );
+    $this->render_field_description(
+      __( 'Create new users in the WordPress database when signups are off. ', 'wp-auth0' ) .
```

Comment:
if here you refer to the "auth0 database signups" I'd clarify that:


Create new users in the WordPress database when signups are disabled in the auth0 database connection and an existing auth0 user logs in. Or something like that. It's not clearer my suggestion, sorry haha.

Reply:
Not clear or correct :)

This is ... WordPress registrations are off, user is added to Auth0 and logs in, user is added to WordPress. I think what I have is about as straightforward as I can get (after a few revisions) but open to suggestions.

Reply:
How can users be "added to wordpress" if "wordpress registrations are off"?

Reply:
You add in Auth0, successfully login, then they are added to WP

Reply:
Is that expected? I understand there are 2 different user databases, with duplicated info IMO. What's the point on keeping a "auth0 user" that would never be able to "create a wordpress account"?

Reply:
Looking through how this operates ... the scenario here is that WP registrations are turned off so the Lock form hides the **Signup** tab and the WP registration form is off. This means that no one can register themselves in WP (users can still be added by an admin). Now, an Auth0 admin adds a user in Auth0 and provides them their username and password. If the setting is turned:

- **off** (default) then the login is blocked (`throw new WP_Auth0_RegistrationNotEnabledException();` in `WP_Auth0_UsersRepo::create()`)
- **on** then the user is created as usual and life continues

I'm not saying this is *how* it should work, just how it currently does. IMHO, this should at least be turned on by default, maybe just forced to be the case.